File: server/backend/app/relay/protocol.py
"""Pydantic models for the GLSD WebSocket protocol.

Field names match the JSON wire format defined in node/protocol-go/messages.go.
Use camelCase aliases via Field(alias=...) since the wire format is camelCase.
"""
from __future__ import annotations

import typing

from pydantic import BaseModel, Field


class HelloMessage(BaseModel):
    type: typing.Literal["hello"]
    machine_id: str = Field(alias="machineId")
    daemon_version: str = Field(alias="daemonVersion")
    os: str
    arch: str
    last_sequence_by_session: dict[str, int] = Field(
        default_factory=dict, alias="lastSequenceBySession"
    )
    model_config = {"populate_by_name": True}


class WelcomeMessage(BaseModel):
    type: typing.Literal["welcome"] = "welcome"
    acked_sequences_by_session: dict[str, int] = Field(
        default_factory=dict, alias="ackedSequencesBySession"
    )
    model_config = {"populate_by_name": True}


class AckMessage(BaseModel):
    type: typing.Literal["ack"] = "ack"
    session_id: str = Field(alias="sessionId")
    sequence_number: int = Field(alias="sequenceNumber")
    model_config = {"populate_by_name": True}


class TaskMessage(BaseModel):
    type: typing.Literal["task"]
    task_id: str = Field(alias="taskId")
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    prompt: str
    model: str
    effort: str
    permission_mode: str = Field(alias="permissionMode")
    persona_system_prompt: str = Field(default="", alias="personaSystemPrompt")
    cwd: str
    claude_session_id: str = Field(default="", alias="claudeSessionId")
    model_config = {"populate_by_name": True}


class StopMessage(BaseModel):
    type: typing.Literal["stop"]
    channel_id: str = Field(alias="channelId")
    session_id: str = Field(alias="sessionId")
    model_config = {"populate_by_name": True}


class StreamMessage(BaseModel):
    type: typing.Literal["stream"]
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    sequence_number: int = Field(alias="sequenceNumber")
    event: dict | list | str = Field(default={})
    model_config = {"populate_by_name": True}


class HeartbeatMessage(BaseModel):
    type: typing.Literal["heartbeat"]
    machine_id: str = Field(alias="machineId")
    daemon_version: str = Field(alias="daemonVersion")
    status: str
    timestamp: str
    model_config = {"populate_by_name": True}


class TaskStartedMessage(BaseModel):
    type: typing.Literal["taskStarted"]
    task_id: str = Field(alias="taskId")
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    started_at: str = Field(alias="startedAt")
    model_config = {"populate_by_name": True}


class TaskCompleteMessage(BaseModel):
    type: typing.Literal["taskComplete"]
    task_id: str = Field(alias="taskId")
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    claude_session_id: str = Field(alias="claudeSessionId")
    input_tokens: int = Field(alias="inputTokens")
    output_tokens: int = Field(alias="outputTokens")
    cost_usd: str = Field(alias="costUsd")
    duration_ms: int = Field(alias="durationMs")
    result_summary: str = Field(default="", alias="resultSummary")
    model_config = {"populate_by_name": True}


class TaskErrorMessage(BaseModel):
    type: typing.Literal["taskError"]
    task_id: str = Field(alias="taskId")
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    error: str
    model_config = {"populate_by_name": True}


class PermissionRequestMessage(BaseModel):
    type: typing.Literal["permissionRequest"]
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    request_id: str = Field(alias="requestId")
    tool_name: str = Field(alias="toolName")
    tool_input: dict | list | str = Field(default={}, alias="toolInput")
    model_config = {"populate_by_name": True}


class QuestionMessage(BaseModel):
    type: typing.Literal["question"]
    session_id: str = Field(alias="sessionId")
    channel_id: str = Field(alias="channelId")
    request_id: str = Field(alias="requestId")
    question: str
    options: list[str] = Field(default_factory=list)
    model_config = {"populate_by_name": True}


class PermissionResponseMessage(BaseModel):
    type: typing.Literal["permissionResponse"]
    channel_id: str = Field(alias="channelId")
    session_id: str = Field(alias="sessionId")
    request_id: str = Field(alias="requestId")
    approved: bool
    model_config = {"populate_by_name": True}


class QuestionResponseMessage(BaseModel):
    type: typing.Literal["questionResponse"]
    channel_id: str = Field(alias="channelId")
    session_id: str = Field(alias="sessionId")
    request_id: str = Field(alias="requestId")
    answer: str
    model_config = {"populate_by_name": True}


# Discriminated union for incoming node messages
NodeMessage = (
    HelloMessage
    | StreamMessage
    | HeartbeatMessage
    | TaskStartedMessage
    | TaskCompleteMessage
    | TaskErrorMessage
    | PermissionRequestMessage
    | QuestionMessage
)

# Discriminated union for incoming browser messages
BrowserMessage = (
    TaskMessage | StopMessage | PermissionResponseMessage | QuestionResponseMessage
)

File: server/frontend/src/lib/protocol.ts
// GLSD — TypeScript protocol types mirroring protocol-go/messages.go
// Field names MUST match Go struct json tags exactly (camelCase)
// See: node/protocol-go/messages.go and node/protocol-go/PROTOCOL.md

// ============================================================
// Browser -> Server messages
// ============================================================

export interface TaskMessage {
  type: 'task';
  taskId: string;
  sessionId: string;
  channelId: string;
  prompt: string;
  model: string;
  effort: string;
  permissionMode: string;
  personaSystemPrompt?: string;
  cwd: string;
  claudeSessionId?: string;
}

export interface StopMessage {
  type: 'stop';
  channelId: string;
  sessionId: string;
}

export interface PermissionResponseMessage {
  type: 'permissionResponse';
  channelId: string;
  sessionId: string;
  requestId: string;
  approved: boolean;
}

export interface QuestionResponseMessage {
  type: 'questionResponse';
  channelId: string;
  sessionId: string;
  requestId: string;
  answer: string;
}

export interface BrowseDirMessage {
  type: 'browseDir';
  requestId: string;
  channelId: string;
  machineId: string;
  path: string;
}

export interface ReadFileMessage {
  type: 'readFile';
  requestId: string;
  channelId: string;
  machineId: string;
  path: string;
  maxBytes?: number;
}

// ============================================================
// Server -> Browser messages
// ============================================================

export interface StreamMessage {
  type: 'stream';
  sessionId: string;
  channelId: string;
  sequenceNumber: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  event: any;
}

export interface TaskStartedMessage {
  type: 'taskStarted';
  taskId: string;
  sessionId: string;
  channelId: string;
  startedAt: string;
}

export interface TaskCompleteMessage {
  type: 'taskComplete';
  taskId: string;
  sessionId: string;
  channelId: string;
  claudeSessionId: string;
  inputTokens: number;
  outputTokens: number;
  costUsd: string;
  durationMs: number;
  resultSummary?: string;
}

export interface TaskErrorMessage {
  type: 'taskError';
  taskId: string;
  sessionId: string;
  channelId: string;
  error: string;
}

export interface PermissionRequestMessage {
  type: 'permissionRequest';
  sessionId: string;
  channelId: string;
  requestId: string;
  toolName: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  toolInput: any;
}

export interface QuestionMessage {
  type: 'question';
  sessionId: string;
  channelId: string;
  requestId: string;
  question: string;
  options?: string[];
}

export interface HeartbeatMessage {
  type: 'heartbeat';
  machineId: string;
  daemonVersion: string;
  status: string;
  timestamp: string;
}

export interface BrowseDirResultMessage {
  type: 'browseDirResult';
  requestId: string;
  channelId: string;
  ok: boolean;
  entries?: BrowseEntry[];
  error?: string;
}

export interface ReadFileResultMessage {
  type: 'readFileResult';
  requestId: string;
  channelId: string;
  ok: boolean;
  content?: string;
  truncated?: boolean;
  error?: string;
}

// ============================================================
// Connection handshake messages (daemon <-> server)
// ============================================================

export interface HelloMessage {
  type: 'hello';
  machineId: string;
  daemonVersion: string;
  os: string;
  arch: string;
  lastSequenceBySession: Record<string, number>;
}

export interface WelcomeMessage {
  type: 'welcome';
  ackedSequencesBySession: Record<string, number>;
}

export interface AckMessage {
  type: 'ack';
  sessionId: string;
  sequenceNumber: number;
}

export interface ReplayRequestMessage {
  type: 'replayRequest';
  sessionId: string;
  fromSequence: number;
}

export interface ReplayCompleteMessage {
  type: 'replayComplete';
  sessionId: string;
  lastSequence: number;
}

// ============================================================
// Supporting types
// ============================================================

export interface BrowseEntry {
  name: string;
  path: string;
  isDirectory: boolean;
  size: number;
  modifiedAt: string;
}

// ============================================================
// gsd2Query / gsd2QueryResult — generic daemon command channel
// Browser sends gsd2Query; relay routes by machineId; daemon responds with gsd2QueryResult.
// machineId is consumed by the relay for routing and not forwarded to the daemon.
// ============================================================

export interface Gsd2QueryMessage {
  type: 'gsd2Query';
  requestId: string;
  channelId: string;
  machineId: string;
  command: string;
  params?: Record<string, unknown>;
}

export interface Gsd2QueryResultMessage {
  type: 'gsd2QueryResult';
  requestId: string;
  channelId: string;
  ok: boolean;
  data?: unknown;
  error?: string;
}

export function isGsd2QueryResult(msg: unknown): msg is Gsd2QueryResultMessage {
  return (
    typeof msg === 'object' &&
    msg !== null &&
    (msg as Gsd2QueryResultMessage).type === 'gsd2QueryResult'
  );
}

// ============================================================
// Discriminated union — all protocol messages
// ============================================================

export type ProtocolMessage =
  | TaskMessage
  | StopMessage
  | PermissionResponseMessage
  | QuestionResponseMessage
  | BrowseDirMessage
  | ReadFileMessage
  | Gsd2QueryMessage
  | StreamMessage
  | TaskStartedMessage
  | TaskCompleteMessage
  | TaskErrorMessage
  | PermissionRequestMessage
  | QuestionMessage
  | HeartbeatMessage
  | BrowseDirResultMessage
  | ReadFileResultMessage
  | Gsd2QueryResultMessage
  | HelloMessage
  | WelcomeMessage
  | AckMessage
  | ReplayRequestMessage
  | ReplayCompleteMessage;

// ============================================================
// Type guard helpers
// ============================================================

export function isStreamMessage(msg: ProtocolMessage): msg is StreamMessage {
  return msg.type === 'stream';
}

export function isTaskStarted(msg: ProtocolMessage): msg is TaskStartedMessage {
  return msg.type === 'taskStarted';
}

export function isTaskComplete(msg: ProtocolMessage): msg is TaskCompleteMessage {
  return msg.type === 'taskComplete';
}

export function isTaskError(msg: ProtocolMessage): msg is TaskErrorMessage {
  return msg.type === 'taskError';
}

export function isPermissionRequest(msg: ProtocolMessage): msg is PermissionRequestMessage {
  return msg.type === 'permissionRequest';
}

export function isQuestion(msg: ProtocolMessage): msg is QuestionMessage {
  return msg.type === 'question';
}

export function isHeartbeat(msg: ProtocolMessage): msg is HeartbeatMessage {
  return msg.type === 'heartbeat';
}

export function isBrowseDirResult(msg: ProtocolMessage): msg is BrowseDirResultMessage {
  return msg.type === 'browseDirResult';
}

export function isReadFileResult(msg: ProtocolMessage): msg is ReadFileResultMessage {
  return msg.type === 'readFileResult';
}

export function isReplayComplete(msg: ProtocolMessage): msg is ReplayCompleteMessage {
  return msg.type === 'replayComplete';
}
